src/FEValues/FunctionValues.jl

```diff
@@ -0,0 +1,184 @@
+
+# Helpers to get the correct types for FunctionValues for the given function and, if needed, geometric interpolations.
+struct SInterpolationDims{rdim,sdim} end
+struct VInterpolationDims{rdim,sdim,vdim} end
+function InterpolationDims(::ScalarInterpolation, ip_geo::VectorizedInterpolation{sdim}) where sdim
+    return SInterpolationDims{getdim(ip_geo),sdim}()
+end
+function InterpolationDims(::VectorInterpolation{vdim}, ip_geo::VectorizedInterpolation{sdim}) where {vdim,sdim}
+    return VInterpolationDims{getdim(ip_geo),sdim,vdim}()
+end
+
+typeof_N(::Type{T}, ::SInterpolationDims) where T = T
+typeof_N(::Type{T}, ::VInterpolationDims{<:Any,dim,dim}) where {T,dim} = Vec{dim,T}
+typeof_N(::Type{T}, ::VInterpolationDims{<:Any,<:Any,vdim}) where {T,vdim} = SVector{vdim,T} # Why not ::Vec here?
+
+typeof_dNdx(::Type{T}, ::SInterpolationDims{dim,dim}) where {T,dim} = Vec{dim,T}
+typeof_dNdx(::Type{T}, ::SInterpolationDims{<:Any,sdim}) where {T,sdim} = SVector{sdim,T} # Why not ::Vec here?
+typeof_dNdx(::Type{T}, ::VInterpolationDims{dim,dim,dim}) where {T,dim} = Tensor{2,dim,T}
+typeof_dNdx(::Type{T}, ::VInterpolationDims{<:Any,sdim,vdim}) where {T,sdim,vdim} = SMatrix{vdim,sdim,T} # If vdim=sdim!=rdim Tensor would be possible...
+
+typeof_dNdξ(::Type{T}, ::SInterpolationDims{dim,dim}) where {T,dim} = Vec{dim,T}
+typeof_dNdξ(::Type{T}, ::SInterpolationDims{rdim}) where {T,rdim} = SVector{rdim,T} # Why not ::Vec here?
+typeof_dNdξ(::Type{T}, ::VInterpolationDims{dim,dim,dim}) where {T,dim} = Tensor{2,dim,T}
+typeof_dNdξ(::Type{T}, ::VInterpolationDims{rdim,<:Any,vdim}) where {T,rdim,vdim} = SMatrix{vdim,rdim,T} # If vdim=rdim!=sdim Tensor would be possible...
+
+"""
+    FunctionValues(::Type{T}, ip_fun, qr::QuadratureRule, ip_geo::VectorizedInterpolation)
+
+Create a `FunctionValues` object containing the shape values and gradients for both the reference 
+cell (precalculated) and the real cell (updated in `reinit!`). 
+"""
+FunctionValues
+
+struct FunctionValues{IP, N_t, dNdx_t, dNdξ_t}
+    ip::IP          # ::Interpolation
+    N_x::N_t        # ::AbstractMatrix{Union{<:Tensor,<:Number}}
+    N_ξ::N_t        # ::AbstractMatrix{Union{<:Tensor,<:Number}}
+    dNdx::dNdx_t    # ::AbstractMatrix{Union{<:Tensor,<:StaticArray}}
+    dNdξ::dNdξ_t    # ::AbstractMatrix{Union{<:Tensor,<:StaticArray}}
+end
+function FunctionValues(::Type{T}, ip::Interpolation, qr::QuadratureRule, ip_geo::VectorizedInterpolation) where T
+    ip_dims = InterpolationDims(ip, ip_geo)
+    N_t = typeof_N(T, ip_dims)
+    dNdx_t = typeof_dNdx(T, ip_dims)
+    dNdξ_t = typeof_dNdξ(T, ip_dims)
+    n_shape = getnbasefunctions(ip)
+    n_qpoints = getnquadpoints(qr)
+    
+    N_ξ  = zeros(N_t,    n_shape, n_qpoints)
+    if isa(get_mapping_type(ip), IdentityMapping)
+        N_x = N_ξ
+    else
+        N_x  = zeros(N_t,    n_shape, n_qpoints)
+    end
+    dNdξ = zeros(dNdξ_t, n_shape, n_qpoints)
+    dNdx = fill(zero(dNdx_t) * T(NaN), n_shape, n_qpoints)
```

Comment:
Perhaps the types above can be inlined here
```suggestion
    dNdx = fill(zero(typeof_dNdx(T, ip_dims)) * T(NaN), n_shape, n_qpoints)
```
and similar